Configuration & Server Overhaul - Tool Shed and Reports

jmchilton opened the pull request:
- Follow up on broken #2837 with an at least less broken attempt bring startup scripts and config processing inline between Galaxy, Reports, and the Tool Shed.
- Define a schema format and tooling (derived from pywalify) to define web server and application option documentation and validation.
- Update Galaxy common startup functions to allow starting up uwsgi.
- Update Galaxy to allow YAML-based configuration files.
- Switch out of the box server for use by tool shed and reports to uwsgi.
- Add schema validation for uwsgi options (derived from uwsgi docs).
- Add validation/documentation schema for tool shed including
  - Utilities to convert of .ini configs to new YAML ones and switch paste to uwsgi.
  - Build a new yaml sample that targets uwsgi out of the box.
  - Utility to validate a tool shed configuration file in the new YAML format.
- Add validation/documentation schema for reports including
  - Utilities to convert of .ini configs to new YAML ones and switch paste to uwsgi.
  - Build a new yaml sample that targets uwsgi out of the box.
  - Utility to validate a reports configuration file in the new YAML format.
- Add reports documentation to the admin section and documentation of the configuration options derived from the schema.

There is no additional duplication between various items (before a .ini.sample had a default and description) now the schema defines these items and everything else (docs, validation, sample YAML) are derived from that. The new validation format should be more usable from within the actual applications themselves to replace existing type conversion and default handling logic - an exciting future direction for this work.

This is hopefully a test bed to bring these concepts quickly to Galaxy - I think all of the infrastructure needed to do this is now readily available after this PR.

jmchilton committed:
Bring Galaxy, reports, and shed run scripts in sync.

- They now uniformly handle server option processing.
- Reports and Tool Shed now respect many new and important options such as --skip_venv.
- Reports and Tool Shed now do extra checking around Python version that Galaxy does.

In addition to improving the reports and tool shed startup scripts and reducing cognative load with respect to option handling, this is an important pre-condition to us being able us to switch on uwsgi by default for all three services simultaneously without a bunch of copy and pasting.

jmchilton committed:
Centralize launching server for reports and tool shed.

Initial outline of allowing launch via uwsgi.

jmchilton committed:
Update tool shed to use newer configuration schema.

 - Implement kwalify schema to validate/define the schema.
 - Replace ini configuration with YAML (completely backward compatible).
 - Rebuild a sample configuration from the schema (config/tool_shed.yml.sample).

jmchilton committed:
Update galaxy.util.properties to allow YAML.

jmchilton committed:
Scripts to manage configuration infrastructure.

jmchilton committed:
Makefile shortcuts for configuration script.

jmchilton committed:
Add uwsgi_schema.yml.

jmchilton committed:
Update Galaxy Reports to use newer configuration schema.

 - Implement kwalify schema to validate/define the schema.
 - Replace ini configuration with YAML (completely backward compatible).
 - Rebuild a sample configuration from the schema (config/reports.yml.sample).

jmchilton committed:
Add reports documentation to ``doc/``.

- Automatically generate RST docs from config schema.
- Describe older functionality as well as new YAML-based configuration.
- Replace https://wiki.galaxyproject.org/Admin/UsageReports.

jmchilton committed:
Add required dependency on uwsgi.

Required for the new out-of-the-box configuration of tool shed and reports.

jmchilton committed:
Fix for empty app:main during conversion.

martenson commented:
is this ready for review @jmchilton ?

jmchilton commented:
@martenson - Absolutely.

martenson commented on scripts/config_manage.py:
```diff
@@ -0,0 +1,473 @@
+    for section in p.sections():
+        if section.startswith("server:"):
+            if server_section:
+                message = "Additional sever section after [%s] encountered [%s], will be ignored."
```
s/sever/server

martenson commented on Makefile:
```diff
@@ -32,6 +32,30 @@ open-project: ## open project on github
+reports-config-validate: ## validate reports YAML configuration file
+	$(IN_VENV) python scripts/config_manage.py validate reports
+
+reports-config-convert: ## convert old style reports ini to yaml
```
When trying to convert my config using this I get the following output in the console but no new file is created and reports.ini remains the same.

```
$ make reports-config-convert
if [ -f .venv/bin/activate ]; then . .venv/bin/activate; fi; python scripts/config_manage.py convert reports
uwsgi:

  http: 127.0.0.1:9001

  threads: 10

  http-raw-body: True

  offload-threads: 8

reports:

  database_connection: postgres://<usr:pswd>@localhost:5432/galaxy-git
```

martenson replied:
This output above seems to be missing the `module: galaxy.webapps.reports.buildapp:uwsgi_app()` line which is in the yml.sample however even if I add it I am getting the following error when trying to use the yml config as reports.ini: 

```
ConfigParser.MissingSectionHeaderError: File contains no section headers.
file: /Users/marten/devel/git/galaxy/config/reports.ini, line: 1
'uwsgi:\n'
```

When I take the whole new config (in yml) and paste it to the reports.yml.sample (removing the reports.ini completely) it also does not work. (With vanilla sample config run the app runs ok.)

martenson commented on scripts/config_manage.py:
```diff
@@ -0,0 +1,473 @@
+def _validate(args, app_desc):
+    path = app_desc.destination
+    if not os.path.exists(path):
+        _warn("Path [%s] not a file, using sample.")
```
this does not seem to succeed for reports sample:

```
$ make reports-config-validate
if [ -f .venv/bin/activate ]; then . .venv/bin/activate; fi; python scripts/config_manage.py validate reports
WARNING: Path [%s] not a file, using sample.
Traceback (most recent call last):
  File "scripts/config_manage.py", line 473, in <module>
    main()
  File "scripts/config_manage.py", line 188, in main
    _validate(args, app_desc)
  File "scripts/config_manage.py", line 283, in _validate
    schema_files=[name],
TypeError: 'NoneType' object is not callable
make: *** [reports-config-validate] Error 1
```

martenson commented on scripts/config_manage.py:
```diff
@@ -0,0 +1,473 @@
+
+
+def _validate(args, app_desc):
+    path = app_desc.destination
```
this does not seem to correctly read the app_desc since it does not find `config/reports.ini` when running `$ make reports-config-validate`

jmchilton replied:
Validate currently only works on the YAML files, I'll see if I can make that more clear.

martenson replied:
It was a yaml file. Do you mean that it has to end `.yml`?

martenson reviewed (changes requested):
This is feedback on the Reports webapp part only. I like the way this is heading; please see few problems I have encountered inline. 

General question: do you expect trouble converting deployers' old configurations into the yaml format? Do you expect to do this migration by default in some future release?
Thanks for working on this @jmchilton

jmchilton commented:
> Do you expect trouble converting deployers' old configurations into the yaml format?

I wouldn't expect it to be perfect, but I think we can do a good job.

> Do you expect to do this migration by default in some future release?

Definitely not.

martenson commented:
@jmchilton Do you want me to review the TS part at the moment or should I wait for the resolution of the reports feedback?

jmchilton commented:
@martenson As long as you have no ... ... big picture objections to anything I'll dig back in and keep working on this this afternoon and hopefully have an update. If you want to wait for that - it would seem totally fair. There is some pressure to get this done to increase the robustness of the GIEs.

jmchilton committed:
Spelling fix caught by @martenson.

martenson commented:
@jmchilton I will review again tomorrow then, focusing on the TS side.

jmchilton committed:
Have convert Makefile tasks actually write outputs.

During @martenson's review he noted that the convert options just print new config files to the screen they don't actually write config options.

- Add an option to the config_manage.py for doing a ``--dry-run``.
- If ``--dry-run`` isn't flagged, actually backup the original file and write the new one.
- Add new Makefile targets for --dry-run conversion.
- Include output about what is handling during the conversion process.
- Revise command-line arg handling a bit.

The new user experience:

```
$ make reports-config-convert-dry-run
if [ -f .venv/bin/activate ]; then . .venv/bin/activate; fi; python scripts/config_manage.py convert reports --dry-run
Writing the file contents:
 |uwsgi:
 |
 |  http: 127.0.0.1:9001
 |
 |  threads: 10
 |
 |  http-raw-body: True
 |
 |  offload-threads: 8
 |
 |reports:
 |
to config/reports.yml
... skipping because --dry-run is enabled.
Moving [config/reports.ini] to [config/reports.ini.backup]
... skipping because --dry-run is enabled.
$ make reports-config-convert
if [ -f .venv/bin/activate ]; then . .venv/bin/activate; fi; python scripts/config_manage.py convert reports
Writing the file contents:
 |uwsgi:
 |
 |  http: 127.0.0.1:9001
 |
 |  threads: 10
 |
 |  http-raw-body: True
 |
 |  offload-threads: 8
 |
 |reports:
 |
to config/reports.yml
Moving [config/reports.ini] to [config/reports.ini.backup]
```

jmchilton committed:
Fixes for two error messages reported by @martenson.

One warning had an unexpanded template variable, the other error message was obsecured. Validation requires pykwalify, the error message now valiates this:

```
if [ -f .venv/bin/activate ]; then . .venv/bin/activate; fi; python scripts/config_manage.py validate reports
WARNING: Path [config/reports.yml] not a file, using sample.
Traceback (most recent call last):
  File "scripts/config_manage.py", line 505, in <module>
    main()
  File "scripts/config_manage.py", line 187, in main
    _validate(args, app_desc)
  File "scripts/config_manage.py", line 294, in _validate
    raise Exception("Cannot validate file, pykawlify is not installed.")
Exception: Cannot validate file, pykawlify is not installed.
make: *** [reports-config-validate] Error 1
```

jmchilton committed:
Add pykwalify as dependency for the new config validation stuff.

xref https://github.com/galaxyproject/starforge/pull/110

jmchilton committed:
Spelling fix thanks to @martenson.

jmchilton committed:
Fix a few bugs in the config conversion process.

- Add module to uwsgi configuration.
- Emulate proper keys and values for lines of the YAML dict.

jmchilton committed:
Move config_manage into lib.

I think we will want to reuse some abstractions and it should make unit testing a bit clearer.

jmchilton commented:
@martenson I made some good progress addressing your issues yesterday but today I'm working on setting up unit tests for all of this - might be best to wait another day before reviewing again.

jmchilton committed:
Use Galaxy's nicer config parsing in config_manage...

So that %(here) works.

jmchilton committed:
Add infrastructure for unit testing config_manage.

martenson commented:
🎷 
ping me when you need me

jmchilton committed:
Improve structure of config_mange with dict.

jmchilton committed:
More test cases and more refactoring of config_manage.

jmchilton commented:
Closing this until I can schedule some time with @natefoo to go over it.

jmchilton commented:
@natefoo gave me some advice on prefixing with uwsgi:
- http://stackoverflow.com/questions/11292217/run-django-app-via-nginxuwsgi-in-a-subpath
- http://uwsgi-docs.readthedocs.io/en/latest/Nginx.html#hosting-multiple-apps-in-the-same-process-aka-managing-script-name-and-path-info

jmchilton commented:
## Prefix with an nginx proxy

Actually seems simpler than before. All needed in nginx.conf is:

```
    upstream shed_app {
        server 127.0.0.1:9009;
    }

   ...

   server {
        ...
        location /shed {
            uwsgi_pass shed_app;
            include uwsgi_params;
        }
        ...
   }
```

tool_shed.yml requires following fields:

```
uwsgi:
  socket: 127.0.0.1:9009
  mount: /shed=galaxy.webapps.tool_shed.buildapp:uwsgi_app()
  manage-script-name: true
```

# Prefix without nginx proxy:

```
uwsgi:
  http: 127.0.0.1:9009
  mount: /shed=galaxy.webapps.tool_shed.buildapp:uwsgi_app()
  manage-script-name: true
```

martenson commented:
@jmchilton do you want to re-open this and push for a release inclusion?

jmchilton commented:
@martenson Nope - no rush on my part. Happy with this being the "2017" way to ~~deploy~~ _configure_ Galaxy :smile:.

martenson commented:
I thought 2017 is the year of `$ apt-get install galaxy` 😉

jmchilton commented:
It will be - I updated my comment with more precise language.

martenson commented:
@jmchilton updating comments to make me look silly!

natefoo commented:
+:100:

jmchilton commented:
Proxy prefix handling now handled in https://github.com/jmchilton/galaxy/commit/a4132f054246fb7f192cfac9d3eddda7ff96dacd - linking the individual commit since it will likely be rebased into larger commit before next PR.
